// @naxmefy/jdbc/lib/Connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Promise = require("bluebird");
const Statement_1 = require("./Statement");
const PreparedStatement_1 = require("./PreparedStatement");
const CallableStatement_1 = require("./CallableStatement");
class Connection {
    constructor(connection) {
        this.connection = Promise.promisifyAll(connection);
    }
    setAutoCommit(autoCommit) {
        return this.connection.setAutoCommitSync(autoCommit);
    }
    createStatement() {
        return this.connection.createStatementAsync()
            .then((statement) => new Statement_1.Statement(statement));
    }
    prepareStatement(sql) {
        return this.connection.prepareStatementAsync(sql)
            .then((prepareStatement) => new PreparedStatement_1.PreparedStatement(prepareStatement));
    }
    prepareCall(call) {
        return this.connection.prepareCallAsync(call)
            .then((callableStatement) => new CallableStatement_1.CallableStatement(callableStatement));
    }
    commit() {
        return this.connection.commitAsync();
    }
    rollback() {
        return this.connection.rollbackAsync();
    }
    close() {
        return this.connection.closeAsync();
    }
    isClosed() {
        return this.connection.isClosedSync();
    }
    isValid(timeout) {
        return this.connection.isValidSync(timeout);
    }
}
exports.Connection = Connection;
