// @mykeels/ng2-smart-table/esm2015/lib/lib/data-source/server/server.data-source.js
import { HttpParams } from '@angular/common/http';
import { LocalDataSource } from '../local/local.data-source';
import { ServerSourceConf } from './server-source.conf';
import { getDeepFromObject } from '../../helpers';
import { map, finalize } from 'rxjs/operators';
import { Subject } from 'rxjs';
export class ServerDataSource extends LocalDataSource {
    constructor(http, conf = {}) {
        super();
        this.http = http;
        this.lastRequestCount = 0;
        this.onRequestStartSource = new Subject();
        this.onRequestEndSource = new Subject();
        this.onRequestCompleteSource = new Subject();
        this.onRequestErrorSource = new Subject();
        this.conf = new ServerSourceConf(conf);
        if (!this.conf.endPoint) {
            throw new Error('At least endPoint must be specified as a configuration of the server data source.');
        }
    }
    count() {
        return this.lastRequestCount;
    }
    getElements() {
        return this.requestElements()
            .pipe(map(res => {
            this.lastRequestCount = this.extractTotalFromResponse(res);
            this.data = this.extractDataFromResponse(res);
            return this.data;
        })).toPromise();
    }
    /**
     * Extracts array of data from server response
     * @param res
     * @returns {any}
     */
    extractDataFromResponse(res) {
        const rawData = res.body;
        const data = !!this.conf.dataKey ? getDeepFromObject(rawData, this.conf.dataKey, []) : rawData;
        if (data instanceof Array) {
            return data;
        }
        throw new Error(`Data must be an array.
    Please check that data extracted from the server response by the key '${this.conf.dataKey}' exists and is array.`);
    }
    /**
     * Extracts total rows count from the server response
     * Looks for the count in the heders first, then in the response body
     * @param res
     * @returns {any}
     */
    extractTotalFromResponse(res) {
        if (res.headers.has(this.conf.totalKey)) {
            return +res.headers.get(this.conf.totalKey);
        }
        else {
            const rawData = res.body;
            return getDeepFromObject(rawData, this.conf.totalKey, 0);
        }
    }
    requestElements() {
        let httpParams = this.createRequesParams();
        this.emitOnRequestStart(httpParams);
        return this.http.get(this.conf.endPoint, { params: httpParams, observe: 'response' }).pipe(finalize(() => this.emitOnRequestEnd())).pipe(map(res => {
            this.emitOnRequestComplete(res);
            return res;
        }, err => {
            this.emitOnRequestError(err);
            return err;
        }));
    }
    onRequestStart() {
        return this.onRequestStartSource.asObservable();
    }
    onRequestEnd() {
        return this.onRequestEndSource.asObservable();
    }
    onRequestComplete() {
        return this.onRequestCompleteSource.asObservable();
    }
    onRequestError() {
        return this.onRequestErrorSource.asObservable();
    }
    emitOnRequestStart(element) {
        this.onRequestStartSource.next(element);
    }
    emitOnRequestEnd() {
        this.onRequestEndSource.next();
    }
    emitOnRequestComplete(element) {
        this.onRequestCompleteSource.next(element);
    }
    emitOnRequestError(element) {
        this.onRequestErrorSource.next(element);
    }
    createRequesParams() {
        let httpParams = new HttpParams();
        httpParams = this.addSortRequestParams(httpParams);
        httpParams = this.addFilterRequestParams(httpParams);
        return this.addPagerRequestParams(httpParams);
    }
    addSortRequestParams(httpParams) {
        if (this.sortConf) {
            this.sortConf.forEach((fieldConf) => {
                httpParams = httpParams.set(this.conf.sortFieldKey, fieldConf.field);
                httpParams = httpParams.set(this.conf.sortDirKey, fieldConf.direction.toUpperCase());
            });
        }
        return httpParams;
    }
    addFilterRequestParams(httpParams) {
        if (this.filterConf.filters) {
            this.filterConf.filters.forEach((fieldConf) => {
                if (fieldConf['search']) {
                    httpParams = httpParams.set(this.conf.filterFieldKey.replace('#field#', fieldConf['field']), fieldConf['search']);
                }
            });
        }
        return httpParams;
    }
    addPagerRequestParams(httpParams) {
        if (this.pagingConf && this.pagingConf['page'] && this.pagingConf['perPage']) {
            httpParams = httpParams.set(this.conf.pagerPageKey, this.pagingConf['page']);
            httpParams = httpParams.set(this.conf.pagerLimitKey, this.pagingConf['perPage']);
        }
        return httpParams;
    }
}
